contracts/Controller.sol

```diff
@@ -122,57 +130,155 @@ contract Controller is Ownable {
-    // }
+    function getBatch(address _otoken) external view returns (bytes32) {
+        OtokenInterface otoken = OtokenInterface(_otoken);
+
+        return
+            keccak256(
+                abi.encode(
+                    otoken.underlyingAsset(),
+                    otoken.strikeAsset(),
+                    otoken.collateralAsset(),
+                    otoken.expiryTimestamp()
+                )
+            );
+    }
 
     /**
-     * @notice Checks if the sender is the operator of the owner’s account
-     * @param _owner The owner of the account
-     * @param _operator account operator
-     * @return true if it is an account operator, otherwise false
+     * @notice get number of vaults in a specific account
+     * @param _accountOwner account owner address
+     * @return number of vaults
      */
-    // function isOperator(address _owner, address _operator) public view returns (bool) {
-    // }
+    function getAccountVaultCounter(address _accountOwner) external view returns (uint256) {
+        return accountVaultCounter[_accountOwner];
+    }
+
+    /**
+     * @notice function to check if otoken is expired
+     * @param _otoken otoken address
+     * @return true if otoken is expired, else return false
+     */
+    function isExpired(address _otoken) public view returns (bool) {
+        uint256 otokenExpiryTimestamp = OtokenInterface(_otoken).expiryTimestamp();
+
+        return now > otokenExpiryTimestamp;
+    }
+
+    /**
+     * @notice Return a specific vault.
+     * @param _owner account owner.
+     * @param _vaultId vault.
+     * @return Vault struct
+     */
+    function getVault(address _owner, uint256 _vaultId) public view returns (MarginAccount.Vault memory) {
+        return vaults[_owner][_vaultId];
+    }
 
     /**
      * @notice Execute actions on a certain vault
      * @dev For each action in the action Array, run the corresponding action
      * @param _actions An array of type Actions.ActionArgs[] which expresses which actions the user want to execute.
-     * @return Vault strcut. The new vault data that has been modified (or null vault if no action affected any vault)
      */
-    // function _runActions(Actions.ActionArgs[] memory _actions) internal returns (MarginAccount.Vault memory) {
-    // }
+    function _runActions(Actions.ActionArgs[] memory _actions) internal {
+        for (uint256 i = 0; i < _actions.length; i++) {
+            Actions.ActionArgs memory action = _actions[i];
+            Actions.ActionType actionType = action.actionType;
+
+            uint256 prevActionVaultId;
+
+            if (actionType == Actions.ActionType.OpenVault) {
+                // check if this action is manipulating the same vault as all other actions, other than SettleVault
+                require(
+                    prevActionVaultId == 0 || action.vaultId == prevActionVaultId,
```

Comment:
hmm what if the prevActionVaultId is 0? So if someone passes an action on vault 0 and an action on vault 1, won't this condition pass?

Maybe there should be a bool of isVaultIdAssigned?

Reply:
Yeah, so the old implementation make sense now
```
 function _checkActionVault(
        uint256 _prevActionVaultId,
        uint256 _currActionVaultId,
        bool _isActionVaultStored
    ) internal returns (uint256, bool) {
        if (_isActionVaultStored) {
            require(_prevActionVaultId == _currActionVaultId, "Controller: can not run actions on different vaults");
        }
        return (_currActionVaultId, true);
    }
```
@antoncoding thoughts ?

Reply:
updated this.